ui/app/controllers/settings/tokens.js

```diff
@@ -48,31 +49,83 @@ export default class Tokens extends Controller {
     return this.store.peekAll('auth-method');
   }
 
+  get hasJWTAuthMethods() {
+    return this.authMethods.any((method) => method.type === 'JWT');
+  }
+
+  get nonTokenAuthMethods() {
+    return this.authMethods.rejectBy('type', 'JWT');
+  }
+
   @action
-  verifyToken() {
+  async verifyToken() {
     const { secret } = this;
-    this.clearTokenProperties();
     const TokenAdapter = getOwner(this).lookup('adapter:token');
 
-    this.set('token.secret', secret);
-    this.set('secret', null);
+    // const isJWT = secret.length > 36 && secret.split('.').length === 3; // TODO: TEMP, HACKY HACK
+    // use regex to determine if "secret" is a JWT
+    const isJWT =
+      secret.length > 36 &&
+      secret.match(/^[a-zA-Z0-9-_]+\.[a-zA-Z0-9-_]+\.[a-zA-Z0-9-_]+$/);
+
+    if (isJWT) {
+      // TODO: Discussion: There may be multiple JWT methods sometime in the future.
+      // A quick assumption would be to just use the first one, but that
+      // won't hold up forever. We should probably have a way to select which one
+      // to use, or at least a way to know which one was used to generate the token.
+      const methodName = await this.authMethods.findBy('type', 'JWT')?.name;
+
+      // If user passes a JWT token, but there is no JWT auth method, throw an error
+      if (!methodName) {
+        this.set('token.secret', undefined);
+        this.signInStatus = 'jwtFailure';
+        return;
+      }
 
-    TokenAdapter.findSelf().then(
-      () => {
-        // Clear out all data to ensure only data the new token is privileged to see is shown
-        this.resetStore();
+      this.clearTokenProperties();
 
-        // Refetch the token and associated policies
-        this.get('token.fetchSelfTokenAndPolicies').perform().catch();
+      // Set bearer token instead of findSelf etc.
+      TokenAdapter.loginJWT(secret, methodName).then(
+        (token) => {
+          this.set('token.secret', token.secret);
```

Comment:
`question`: Should this be:

```
this.token.setProperties({ secret: token.secret, tokenNotFound: false}) // see line 100
```

Reply:
My fear in changing this was that there was something to resetStore() or fetchSelfTokenAndPolicies that would necessitate them taking place before setting tokenNotFound below. I'll test it out now for code consolidation's sake.